design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+creating a trusted source of identity for `CertificateRequest`s, a policy engine
+is able to determine whether that identity is allowed to request that
+certificate, given some policy configuration setup by a cluster administrator.
+
+Although auditing exists in Kubernetes and exposes the identity of the
+requester, its configuration is not always exposed to end users, such as
+when using managed Kubernetes (GKE, EKS etc). In scenarios where
+configuring auditing is available, it is often not preferable and an
+anti-pattern to make runtime decisions on historical audit logs.
+
+The upstream [Kubernetes certificates
+`CertificateSigningRequest`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#certificatesigningrequest-v1-certificates-k8s-io)
+API has identity as part of the resource. As the project intends to align and
+likely make this resource compatible with cert-manager in some way, the
+`CertificateRequest` should expose identical identity details. This means
+extensions or additions to cert-manager (policy, auditing) can have the same
+guarantees about what information is available from both resources.
+
+
+### Goals
+
+- Introduce identity to the `CertificateRequest` resource which is securely
+    enforced by cert-manager
+- Ensure the identity exposed in `CertificateRequest`s has parity with the
+    upstream Kubernetes `CertificateSigningRequest` resource
+- Enable external tooling or additions to cert-manager that can make use of this
+    identity in a useful way
+
+### Non-Goals
+
+- Discuss how this identity can be used (policy, auditing)
+- Make changes to upstream Kubernetes to implement identiy in cert-manager
+- Although considered below, "passed down" identity is not part of this design
+
+## Proposal
+
+Kubernetes does [not currently support immutable
+fields](https://github.com/kubernetes/enhancements/blob/8b9b994136371f1bc938aabf012f4c45535d684c/keps/sig-api-machinery/20190603-immutable-fields.md)
+for CRDs. The cert-manager webhook will be responsible for populating and
+enforcing identity fields which are present on the spec of `CertificateRequest`
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
+    [UserInfo](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#userinfo-v1-authentication-k8s-io)
+
+The webhook will also responsible for enforcing the following during an UPDATE
+operation:
+- No changes to identity fields are allowed to be made
+
+### API Changes
+
+In order to expose the identity of who created `CertificateRequest` resources,
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
```

Comment:
nit: this is actually on the CertificateRequestSpec - it may be worthwhile adding:

```
type CertificateRequestSpec {
    // existing fields
    ...

    // new fields here
}
```
below to make this clearer :)